### PR TITLE
Azure Notification Hubs iOS SDK Preview 2

### PR DESCRIPTION
**Pull request by mpodwysocki**
This pull request finalizes the Azure Notification Hubs for iOS SDK v3 Preview 2. 

Included in this release is the following:
- Installation Template Support
- Refactor tags from `MSInstallation` and `MSInstallationTemplate` to `MSTaggable` protocol
- Update to replace deprecated APIs for new `MS` prefixes
- Update to old `SB` code to ignore deprecation warnings

### CODE REVIEW

File: src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallationTemplate.m [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+- (BOOL)removeTag:(NSString *)tag {
+    return [self removeTags:[NSArray arrayWithObject:tag]];
+}
+
+- (BOOL)removeTags:(NSArray<NSString *> *)tagsToRemove {
+    NSMutableSet *tmpTags = [NSMutableSet setWithSet:self.tags];
+
+    [tmpTags minusSet:[NSSet setWithArray:tagsToRemove]];
```
Comment (brannon) [inline src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallationTemplate.m:68]: Using `minusSet` doesn't allow us to track if the set was modified. If we are going to return `BOOL`, then we should try to make it accurate. That probably means we need to iterate over the tags to remove, check if the tag exists, remove it, and mark the set as "modified".

File: src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallationTemplate.m [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+#pragma mark Tags
+
+- (BOOL)addTag:(NSString *)tag {
+    return [self addTags:[NSArray arrayWithObject:tag]];
+}
+
+- (BOOL)addTags:(NSArray<NSString *> *)tagsToAdd {
+    NSMutableSet *tmpTags = [NSMutableSet setWithSet:self.tags];
+
+    for (NSString *tag in tagsToAdd) {
```
Comment (brannon) [inline src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallationTemplate.m:48]: This doesn't _really_ track changes to the underlying set. If we are going to return `BOOL`, then we should try to make it accurate. That probably means we need to iterate over the tags to add, check if the tag exists, add it, and mark the set as "modified".

File: src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallationTemplate.m [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+#pragma mark Headers
+
+- (void)setHeader:(NSString *)value forKey:(NSString *)key {
+    NSMutableDictionary *tmpHeaders = [NSMutableDictionary dictionaryWithDictionary:headers];
+    [tmpHeaders setObject:value forKey:key];
+    headers = [tmpHeaders copy];
+}
+
+- (void)removeHeader:(NSString *)key {
+       NSMutableDictionary *tmpHeaders = [NSMutableDictionary dictionaryWithDictionary:headers];
```
Comment (brannon) [inline src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallationTemplate.m:87]: too much indentation.

File: src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallationTemplate.m [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+- (void)encodeWithCoder:(nonnull NSCoder *)coder {
+    [coder encodeObject:body forKey:@"body"];
+    [coder encodeObject:tags forKey:@"tags"];
+    [coder encodeObject:headers forKey:@"headers"];
+}
+
+- (NSDictionary *)toDictionary {
+    return [NSDictionary
+        dictionaryWithObjectsAndKeys:body, @"body", [NSArray arrayWithArray:[tags allObjects]], @"tags", headers, @"headers", nil];
```
Comment (brannon) [inline src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallationTemplate.m:130]: Isn't this basically the same as:
```
return @{
    @"body": body,
    @"tags": [tags allObjects],
    @"headers": headers,
};
```
The literal version is much easier to read.

File: src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallationTemplate.h [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+#import <Foundation/Foundation.h>
+#import "MSTaggable.h"
+
+@interface MSInstallationTemplate : NSObject<MSTaggable>
+
+@property(nonatomic, copy) NSString *body;
+@property(nonatomic, copy, readonly) NSDictionary<NSString *, NSString *> *headers;
+
+// Headers
+- (void)setHeader:(NSString *)value forKey:(NSString *)key;
```
[the remark this answers is not in the document]
Comment (mpodwysocki) [reply src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallationTemplate.h:14]: I think the following makes enough sense:
```objc
- (void)setHeaderValue:(NSString *)value forKey:(NSString *)key;
- (void)removeHeaderForKey:(NSString *)key;
- (NSString *)getHeaderForKey:(NSString *)key;
```

File: src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallationTemplate.m (hunk 1 above)
Comment (mpodwysocki) [reply src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallationTemplate.m:68]: Adding change tracking via `isDirty` and KVO

File: src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallation.h [hunk 6]
```diff
@@ -16,9 +20,8 @@
 
 - (NSData *)toJsonData;
 
-- (BOOL)addTags:(NSArray<NSString *> *)tags;
-- (BOOL)removeTags:(NSArray<NSString *> *)tags;
-- (NSArray<NSString *> *)getTags;
-- (void)clearTags;
+- (BOOL)addTemplate:(MSInstallationTemplate *)template forKey:(NSString *)key;
+- (BOOL)removeTemplate:(NSString *)key;
+- (MSInstallationTemplate *)getTemplate:(NSString *)key;
```
[the remark this answers is not in the document]
Comment (mpodwysocki) [reply src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallation.h:25]: renamed as per above discussion

File: src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallation.h [hunk 7]
```diff
@@ -16,9 +20,8 @@
 
 - (NSData *)toJsonData;
 
-- (BOOL)addTags:(NSArray<NSString *> *)tags;
-- (BOOL)removeTags:(NSArray<NSString *> *)tags;
-- (NSArray<NSString *> *)getTags;
-- (void)clearTags;
+- (BOOL)addTemplate:(MSInstallationTemplate *)template forKey:(NSString *)key;
+- (BOOL)removeTemplate:(NSString *)key;
```
[the remark this answers is not in the document]
Comment (mpodwysocki) [reply src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallation.h:24]: Renamed as per above discussion

File: src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallationTemplate.m (hunk 2 above)
Comment (mpodwysocki) [reply src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallationTemplate.m:48]: Adding change tracking via `isDirty` and KVO

File: src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallation.m [hunk 8]
```diff
@@ -103,12 +127,40 @@ - (void)clearTags {
...
 
 - (NSUInteger)hash {
-    return [self.installationID hash] ^ [self.pushChannel hash] ^ [self.tags hash];
+    return [self.installationID hash] ^ [self.pushChannel hash] ^ [self.tags hash] ^ [self.templates hash];
+}
+
+- (BOOL)addTemplate:(MSInstallationTemplate *)template forKey:(NSString *)templateKey {
+    NSMutableDictionary<NSString *, MSInstallationTemplate *> *tmpTemplates = [NSMutableDictionary dictionaryWithDictionary:self.templates];
+
+    if ([[tmpTemplates allKeysForObject:template] count] > 0) {
```
[the remark this answers is not in the document]
Comment (mpodwysocki) [reply src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallation.m]: Renaming to `setTemplate` therefore if you want to overwrite you can.

File: src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallation.m [hunk 9]
```diff
@@ -103,12 +127,40 @@ - (void)clearTags {
...
+        return NO;
+    }
+
+    [tmpTemplates setObject:template forKey:templateKey];
+    self.templates = tmpTemplates;
+    return YES;
+}
+
+- (BOOL)removeTemplate:(NSString *)templateKey {
```
[the remark this answers is not in the document]
Comment (mpodwysocki) [reply src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallation.m]: Now checking for whether the template exists

File: src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallationTemplate.m (hunk 4 above)
Comment (mpodwysocki) [reply src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallationTemplate.m]: Changing to easier to read standard

File: src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallationTemplate.m (hunk 3 above)
Comment (mpodwysocki) [reply src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallationTemplate.m:119]: Fixed indentation

**Pull request merged**